src/java/ognl/OgnlRuntime.java

```diff
@@ -899,12 +913,53 @@ public static Object invokeMethod(Object target, Method method, Object[] argsArr
                 }
             }
 
-            result = method.invoke(target, argsArray);
+            result = invokeMethodInJavaSandbox(target, method, argsArray);
         }
 
         return result;
     }
 
+    private static Object invokeMethodInJavaSandbox(Object target, Method method, Object[] argsArray) throws InvocationTargetException, IllegalAccessException {
+        SecurityManager parentSecurityManager = System.getSecurityManager();
+        Policy parentPolicy;
+        try
+        {
+            if (parentSecurityManager != null) {
+                parentSecurityManager.checkPermission(new SecurityPermission("getPolicy"));
+                parentSecurityManager.checkPermission(new SecurityPermission("setPolicy"));
+                parentSecurityManager.checkPermission(new RuntimePermission("setSecurityManager"));
+            }
+            parentPolicy = Policy.getPolicy();
+        } catch (SecurityException ex) {
+            return method.invoke(target, argsArray);
+        }
+
+        synchronized (Policy.class) {
+            try {
+                Policy.setPolicy(new OgnlPolicy(parentPolicy, _permissions));
+            } catch (SecurityException ex) {
+                return method.invoke(target, argsArray);
+            }
+
+            try {
+                System.setSecurityManager(new OgnlSecurityManager(parentSecurityManager));
+            } catch (SecurityException ex) {
+                Policy.setPolicy(parentPolicy);
+                return method.invoke(target, argsArray);
+            }
+
+            try {
+                return method.invoke(target, argsArray);
+            } catch (SecurityException ex) {
+                ex.printStackTrace();
+                throw new IllegalAccessException("Method [" + method + "] cannot be accessed.");
+            } finally {
+                System.setSecurityManager(parentSecurityManager);
+                Policy.setPolicy(parentPolicy);
+            }
```

Comment:
I have no idea what's going on here :(

Reply:
When OGNL's invocation finished with our policy and security manager, then here restore all things to what was before method invocation because here we just finished our sensitive action so we relaxes JVM - security manager and policy are global JVM and affect all threads and processes of current JVM!